drivers/media/i2c/mira220.c

```diff
@@ -0,0 +1,2481 @@
+			MIRA220_IMAGER_RUN_STOP, NULL);
+	if (ret) {
+		dev_err(&client->dev, "Error setting run reg to stop");
+		return ret;
+	}
+
+	/*
+		  * Wait for one frame to make sure sensor is set to
+		  * software standby in V-blank
+		  *
+		  * frame_time = frame length rows * Tline
+		  * Tline = line length / pixel clock (in MHz)
+		  */
+	frame_time = MIRA220_DEFAULT_FRAME_LENGTH *
+		     MIRA220_DEFAULT_LINE_LENGTH / MIRA220_DEFAULT_PIXEL_CLOCK;
+
+	usleep_range(frame_time, frame_time + 1000); //TODO, set to 1 frametime
+
+	return ret;
+}
+
+// Returns the maximum exposure time in row_length (reg value).
+// Calculation is baded on Mira220 datasheet Section 9.2.
+static u32 mira220_calculate_max_exposure_time(u32 vsize, u32 vblank,
+					       u32 row_length)
+{
+	return (vsize + vblank) -
+	       (int)(MIRA220_GLOB_NUM_CLK_CYCLES / row_length);
+}
+
+static int mira220_write_exposure_reg(struct mira220 *mira220, u32 exposure)
+{
+	struct i2c_client *const client = v4l2_get_subdevdata(&mira220->sd);
+	const u32 max_exposure = mira220_calculate_max_exposure_time(
+		mira220->mode->height, mira220->vblank->val,
+		mira220->mode->row_length);
+	u32 ret = 0;
+	u64 readval = 0;
+	int otherval = 0;
+	u32 capped_exposure = exposure;
+
+	if (exposure > max_exposure) {
+		capped_exposure = max_exposure;
+	}
+
+	printk(KERN_INFO
+	       "[MIRA220]: exposure fun width %d, hblank %d, vblank %d, row len %d, ctrl->val %d capped to %d.\n",
+	       mira220->mode->width, mira220->hblank->val, mira220->vblank->val,
+	       mira220->mode->row_length, exposure, capped_exposure);
+
+	/*test cci write*/
+	ret = cci_read(mira220->regmap, MIRA220_EXP_TIME_REG, &readval, &ret);
+	otherval = readval & 0xFFFF;
+	//  printk(KERN_INFO "[MIRA220]: HIGH EXP Read reg %d, otherval= %d\n",
+	// MIRA220_EXP_TIME_REG,otherval);
+	ret = cci_write(mira220->regmap, MIRA220_EXP_TIME_REG, capped_exposure,
+			NULL);
+	/*ret = cci_write(mira220->regmap, MIRA220_EXP_TIME_LO_REG_NEW,
+		0xAA, NULL);
+	 usleep_range(1000000, 1000000+100);
+	 ret = cci_read(mira220->regmap,MIRA220_EXP_TIME_LO_REG_NEW, &readval, NULL );
+	 printk(KERN_INFO "[MIRA220]: LOW EXP Read reg %d, val = %llu.\n",
+		MIRA220_EXP_TIME_LO_REG_NEW, readval);
+	 */
+
+	//ret = mira220_write16(mira220, MIRA220_EXP_TIME_LO_REG, capped_exposure);
+	if (ret) {
+		dev_err_ratelimited(&client->dev,
+				    "Error setting exposure time to %d",
+				    capped_exposure);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+// Gets the format code if supported. Otherwise returns the default format code `codes[0]`
+static u32 mira220_validate_format_code_or_default(struct mira220 *mira220,
+						   u32 code)
+{
+	struct i2c_client *client = v4l2_get_subdevdata(&mira220->sd);
+	unsigned int i;
+
+	lockdep_assert_held(&mira220->mutex);
+
+	for (i = 0; i < ARRAY_SIZE(codes); i++)
+		if (codes[i] == code)
+			break;
+
+	if (i >= ARRAY_SIZE(codes)) {
+		dev_err_ratelimited(&client->dev,
```

Comment:
Error message not needed, and gives a route for userspace to be able to spam the kernel log (even if rate limited).